lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,188 @@
+    -------
+    >>> import streamlit as st
+    >>> from pages import page1, page2
+    >>>
+    >>>	pg = st.navigation([st.Page(page1), st.Page(page2)])
+    >>>
+    >>>	st.title("My Awesome App")
+    >>> pg.run()
+    """
+    ctx = get_script_run_ctx()
+    if not ctx:
+        return None
+
+    nav_sections = {"": pages} if isinstance(pages, list) else pages
+    page_list = pages_from_nav_sections(nav_sections)
+
+    if len(page_list) == 0:
+        raise StreamlitAPIException(
+            "`st.navigation` must be called with at least one `st.Page`."
+        )
+
+    defaults = []
+    pagehash_to_pageinfo: dict[PageHash, PageInfo] = {}
+
+    # This nested loop keeps track of three things:
+    # 1. the default pages
+    # 2. the pagehash to pageinfo mapping
+    for section_header in nav_sections:
+        for page in nav_sections[section_header]:
+            if page.default:
+                defaults.append(page)
+
+            if isinstance(page._page, Path):
+                script_path = str(page._page)
+            else:
+                script_path = ""
+
+            script_hash = page._script_hash
+            if script_hash in pagehash_to_pageinfo:
+                raise StreamlitAPIException(
+                    f"Multiple Pages specified with title {page.title}. "
+                    "Page titles should be unique. If not specified, "
+                    "title is inferred from the file path or callable name."
+                )
+
+            pagehash_to_pageinfo[script_hash] = {
+                "page_script_hash": script_hash,
+                "page_name": page.title,
+                "icon": page.icon,
+                "script_path": script_path,
+                "url_pathname": page.title.replace(" ", "_"),
+            }
+
+    # First assume the first page is the default. We will update this if
+    # we detect that a different page is the default.
+    default_page = page_list[0]
+    if len(defaults) > 1:
+        raise StreamlitAPIException(
+            "Multiple Pages specified with `default=True`. At most one Page can be set to default."
+        )
+    if len(defaults) == 0:
+        default_page.default = True
+    else:
+        default_page = defaults[0]
+
+    msg = ForwardMsg()
+    msg.navigation.position = position
+    msg.navigation.sections[:] = nav_sections.keys()
+    for section_header in nav_sections:
+        for page in nav_sections[section_header]:
+            p = msg.navigation.app_pages.add()
+            p.page_script_hash = page._script_hash
+            p.page_name = page.title
+            p.icon = page.icon
+            p.is_default = page.default
+            p.section_header = section_header
+            p.url_pathname = page.title.replace(" ", "_")
+
+    # Inform our page manager about the set of pages we have
+    ctx.pages_manager.set_pages(pagehash_to_pageinfo)
+    managed_page = ctx.pages_manager.get_page_script(
+        fallback_page_hash=default_page._script_hash
```

Comment:
Yea, perhaps better naming is needed here, but it falls under the weirdness we used for the v1, and is just as complex as v2. We basically search for pages based on:
1. `page_script_hash` - if it exists check using this, and otherwise, rely on the default hash, so it should technically always return a page. **The edge case is when `st.navigation` is called with different set of pages**. We set the `page_script_hash` as the page in the first call, but the subsequent calls is with a new set of hashes. We just rely on the new default page identified.
2. We check by page name - this can return `None` - This is the direct navigation setup where we rely on the url path to identify the page and will present a Page Not Found
3. If neither is specified, we get the fallback, it's essentially the default page.

Reply:
Definitely can write more comments around the call. Perhaps `detect_page_script` or `find_page_script`.